src/extend.cpp

```diff
@@ -136,12 +134,17 @@ sparse_coo extend(const sparse_coo &sparse_rr_mat,
   return (std::make_tuple(i_vec_all, j_vec_all, dists_all));
 }
 
-sparse_coo lower_rank(const sparse_coo &sparse_rr_mat,
-                      const size_t n_samples,
-                      const size_t kNN,
-                      bool reciprocal_only,
-                      bool count_unique_distances,
-                      size_t num_threads = 1) {
+struct pair_hash {
+  std::size_t operator()(const std::pair<long,long> & v) const {
+    long hash_value = v.first;
+    hash_value ^= v.second + 0x9e3779b9 + (v.first << 6) + (v.first >> 2);
+    return hash_value;
+  }
+};
```

Comment:
Custom hash? Fair play

Reply:
Note for future testing - we can replace this hash function with an XOR, which is faster but suffers from (a) XOR(i,j) = XOR(j,i) - but not a problem here because we only hash j>i; (b) XOR(i,i) = 0, XOR(j,j) = 0 - but no self loops in the tuple list. No idea if performance improvement but noting it here so I can clear it from my mind.